### PR TITLE
Shared machine component

### PR DESCRIPTION
**Pull request by huwshimi**

### CODE REVIEW

File: jujugui/static/gui/src/app/components/machine-view/machine-unit/machine-unit.js [hunk 1]
```diff
@@ -54,54 +45,38 @@ MachineViewMachineUnitGlobals.collect = function(connect, monitor) {
...
+    const classes = {
+      'machine-view__machine-unit': true,
       'machine-view__machine-unit--draggable': this.props.canDrag,
       'machine-view__machine-unit--dragged': this.props.isDragging
     };
-    classes['machine-view__machine-unit--' + status] = true;
-    return classNames(
-      'machine-view__machine-unit',
-      classes);
+    return Object.keys(classes).filter(className => classes[className]);
```
[the remark this answers is not in the document]
Comment (huwshimi) [reply jujugui/static/gui/src/app/components/machine-view/machine-unit/machine-unit.js]: This is so we can pass in an array of classes.

File: jujugui/static/gui/src/app/components/machine-view/machine-unit/machine-unit.js [hunk 2]
```diff
@@ -54,54 +45,38 @@ MachineViewMachineUnitGlobals.collect = function(connect, monitor) {
...
-          <img
-            alt={unit.displayName}
-            className="machine-view__machine-unit-icon-img"
-            src={service.get('icon')}
-            title={unit.displayName} />
-        </span>
-        {title}
-        {menu}
-      </li>
+      <div>
```
[the remark this answers is not in the document]
Comment (huwshimi) [reply jujugui/static/gui/src/app/components/machine-view/machine-unit/machine-unit.js]: `connectDragSource` requires a div not a component at the top level.

File: jujugui/static/gui/src/app/components/machine-view/machine/machine.js [hunk 3]
```diff
@@ -279,25 +248,28 @@ class MachineViewMachine extends React.Component {
...
-          {this.props.machineAPI.machine.displayName}
-        </div>
-        {this._generateTerminalButton()}
-        {this._generateHardware()}
-        {this._generateUnits()}
-        {this._generateConstraintsForm()}
-        <div className="machine-view__machine-drop-target">
-          <div className="machine-view__machine-drop-message">
-            Add to {this.props.machineAPI.machine.displayName}
+      <div>
```
[the remark this answers is not in the document]
Comment (huwshimi) [reply jujugui/static/gui/src/app/components/machine-view/machine/machine.js]: `connectDropTarget` also requires a div at the top level.

File: jujugui/static/gui/src/app/components/shared/machine/machine.js [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+
+Machine.propTypes = {
+  children: PropTypes.any,
+  classes: PropTypes.arrayOf(PropTypes.string),
+  hardware: PropTypes.arrayOf(PropTypes.shape({
+    label: PropTypes.string.isRequired,
+    value: PropTypes.number.isRequired
+  })),
+  isContainer: PropTypes.bool,
+  machine: PropTypes.shape({
```
[the remark this answers is not in the document]
Comment (huwshimi) [reply jujugui/static/gui/src/app/components/shared/machine/machine.js]: Maybe I've misunderstood you, but they can just pass in an object of that shape (you don't have to create a shape()).

[the remark this answers is not in the document]
Comment (hatched) [reply jujugui/static/gui/src/app/components/shared/machine/machine.js]: Sorry I was quick to comment, I thought I read shapeup

[the remark this answers is not in the document]
Comment (huwshimi) [reply jujugui/static/gui/src/app/components/shared/machine/machine.js]: Ah yeah, no, this is just regular shapes.

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-gce.jujugui.org:8080//job/juju-gui/8916/

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-gce.jujugui.org:8080//job/juju-gui/8922/

**Comment by huwshimi** [comment]
QA:
- add an app and a few units
- open the machine view
- add some machines
- click around the machines, try dropping units on them etc.
- all should be as before except with the new designs
- reload the gui with the `expert` and `terminal` flags on
- connect to an existing model with machines and interact with them in the same ways
- click on the ip address on a machine and the terminal should open

**Comment by huwshimi** [comment]
This is now ready for a review.

File: package-lock.json [hunk 5]
```diff
@@ -3269,7 +3269,8 @@
         "console-control-strings": {
           "version": "1.1.0",
           "bundled": true,
-          "dev": true
+          "dev": true,
```
Comment (hatched) [inline package-lock.json]: Are you using the latest NPM here? I don't get the same changes when I install it.

Comment (huwshimi) [reply package-lock.json]: I've updated using the latest npm. I can't keep up!

**Comment by huwshimi** [comment]
@hatched so there are shared components which I've tried to keep as much GUI specific code out of and then then GUI wrapper components that add the drag and drop ability and gather/pass the correct data to the shared component.

**Comment by hatched** [comment]
Ahh right of course

**Comment by hatched** [comment]
:shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci-gce.jujugui.org:8080/job/juju-gui-merge

**Pull request merged**